### PR TITLE
Fixed #22394 -- Refactored built-in datetime lookups to transforms.

### PR DESCRIPTION
**Pull request by jdufresne**
https://code.djangoproject.com/ticket/22394

### CODE REVIEW

**Comment by jarshwah** [comment]
After a quick look over it looks good to me. I'm not sure that putting the transforms and lookups into `fields/__init__.py` is the right place though. I think I'd prefer them in `db.models.lookups` or in `db.models.fields.lookups`. 

I'll try to review this properly within the next week.

**Comment by jdufresne** [comment]
> I'm not sure that putting the transforms and lookups into fields/__init__.py is the right place though. I think I'd prefer them in db.models.lookups or in db.models.fields.lookups. 

Placing them in `fields/lookups.py` would create a circular import. `fields/lookups.py` is already imported by `fields/__init__.py` but, with that change, `db.models.fields.lookups` would also need to import from `fields/__init__.py`.

There is discussion of the long term organization of the `fields` and `lookups` in PR #4281. Your same concern was raised there. It was agreed to do the re-organization *after* the new lookups went in.

File: django/db/models/fields/__init__.py [hunk 1]
```diff
@@ -2408,3 +2386,150 @@ def formfield(self, **kwargs):
...
+            bounds = connection.ops.year_lookup_bounds_for_date_field(year)
+        return bounds
+
+
+@YearTransform.register_lookup
+class YearExact(YearLookup):
+    lookup_name = 'exact'
+
+    def as_sql(self, compiler, connection):
+        # We will need to skip the extract part, and instead go
```
Comment (timgraham) [inline django/db/models/fields/__init__.py]: chop comma

File: django/db/models/fields/__init__.py [hunk 2]
```diff
@@ -2408,3 +2386,150 @@ def formfield(self, **kwargs):
...
+        rhs_sql = self.get_rhs_op(connection, rhs_sql)
+        start, finish = self.year_lookup_bounds(connection, rhs_params[0])
+        params.append(self.get_bound(start, finish))
+        return '%s %s' % (lhs_sql, rhs_sql), params
+
+    def get_rhs_op(self, connection, rhs):
+        return connection.operators[self.lookup_name] % rhs
+
+    def get_bound(self):
+        raise NotImplementedError
```
Comment (timgraham) [inline django/db/models/fields/__init__.py]: It's nice to include a message for the exception (we had a ticket about adding messages to all of them)

File: django/db/models/fields/__init__.py [hunk 3]
```diff
@@ -2408,3 +2386,150 @@ def formfield(self, **kwargs):
...
         return super(UUIDField, self).formfield(**defaults)
+
+
+class DateTransform(Transform):
+    extract_type = None
+
+    def as_sql(self, compiler, connection):
+        sql, params = compiler.compile(self.lhs)
+        lhs_output_field = self.lhs.output_field
+        if isinstance(lhs_output_field, DateTimeField):
```
Comment (timgraham) [inline django/db/models/fields/__init__.py]: The `isinstance` check feels a bit dirty but I guess it's simpler than having two transform classes. No strong opinion from me, just something to think about.

File: django/db/models/fields/__init__.py [hunk 4]
```diff
@@ -2408,3 +2386,150 @@ def formfield(self, **kwargs):
         }
         defaults.update(kwargs)
         return super(UUIDField, self).formfield(**defaults)
+
+
+class DateTransform(Transform):
+    extract_type = None
```
Comment (timgraham) [inline django/db/models/fields/__init__.py]: could `extract_type` default to `lookup_name`? not sure if it's different for any of the transforms we have so far.

**Comment by jdufresne** [comment]
> The isinstance check feels a bit dirty but I guess it's simpler than having two transform classes. No strong opinion from me, just something to think about.

Totally understand. I thought about this a bit too as I was developing the commit. If you prefer I break this out into more classes I can take that approach as well. But I'll wait for an expressed preference.

> could extract_type default to lookup_name? not sure if it's different for any of the transforms we have so far.

Would you prefer I remove `extract_type` entirely? In the current state of the code, it is never different.

**Comment by jarshwah** [comment]
The only reason I can think of to keep extract_type is so that the lookup_name can be changed and the transform/lookups will still work. I don't think that's going to be very useful though, so we can remove extract_type until it becomes an actual problem.

I think we're OK with the isinstance check for DateTimeField. The cost is small and it helps to simplify the hierarchy. We'd need like 8 extra classes if the split was there which seems unnecessary.

So other than the extract_type it LGTM.

**Comment by jdufresne** [comment]
Thanks everyone. I have removed the `extract_type` as suggested. Any additional feedback is welcome.

**Comment by akaariai** [comment]
We need docs & release notes for this, otherwise a big +1.

**Comment by timgraham** [comment]
Could you clarify what docs are needed? It seemed to me like an internal refactoring with no change in behavior.

**Comment by akaariai** [comment]
The fact you can now do queries like datefield__month__lte is worth documenting. If I'm not mistaken the lookups are documented, but that they can be used as transforms isn't.

**Comment by timgraham** [comment]
Got it. Should add some tests for those cases too?

**Comment by jdufresne** [comment]
I have added docs and tests for the new features of the change. Feedback welcome. Thanks.

**Comment by jarshwah** [comment]
I'm happy with the docs and tests here. Tim?

File: docs/ref/models/querysets.txt [hunk 5]
```diff
@@ -2431,36 +2431,45 @@ numbers and even characters.
...
+    SELECT ... WHERE pub_date >= '2005-01-01'';
 
 (The exact SQL syntax varies for each database engine.)
 
 When :setting:`USE_TZ` is ``True``, datetime fields are converted to the
 current time zone before filtering.
 
+.. versionchanged:: 1.9
+
+   Allow chaining additional field lookups.
```
Comment (timgraham) [inline docs/ref/models/querysets.txt]: 4 space indent
Allow -> Allowed

**Comment by jdufresne** [comment]
Fixed bad formatting. Thanks.

**Comment by timgraham** [comment]
merged in b5e0eede406633b88c6222031171f80876d8f5e1, thanks!

**Pull request closed without merging**